dace/frontend/python/replacements.py

```diff
@@ -814,90 +872,122 @@ def _convert_type(dtype1, dtype2, operator) -> Tuple[dace.dtypes.typeclass]:
                     np.int8, np.int16, np.int32, np.int64}
     # unsigned_types = {np.uint8, np.uint16, np.uint32, np.uint64}
 
-    if dtype1 in complex_types:
-        type1 = 3  # complex
-    elif dtype1 in float_types:
-        type1 = 2  # float
-    elif dtype1 in signed_types:
-        type1 = 1  # signed integer, bool
-    else:
-        type1 = 0  # unsigned integer
-
-    if dtype2 in complex_types:
-        type2 = 3  # complex
-    elif dtype2 in float_types:
-        type2 = 2  # float
-    elif dtype2 in signed_types:
-        type2 = 1  # signed integer, bool
-    else:
-        type2 = 0  # unsigned integer
+    coarse_types = []
+    for dtype in datatypes:
+        if dtype in complex_types:
+            coarse_types.append(3) # complex
+        elif dtype in float_types:
+            coarse_types.append(2)  # float
+        elif dtype in signed_types:
+            coarse_types.append(1) # signed integer, bool
+        else:
+            coarse_types.append(0)  # unsigned integer
+    
+    casting = [None] * len(arguments)
 
-    left_cast = None
-    right_cast = None
+    if len(arguments) == 1:  # Unary operators
 
-    if _is_op_arithmetic(operator):
+        if operator == 'USub' and coarse_types[0] == 0:
+            result_type = eval('dace.int{}'.format(4 * datatypes[0].bytes))
+        elif _is_op_boolean(operator):
+            result_type = dace.bool_
+        else:
+            result_type = datatypes[0]
 
-        # Float division between integers
-        if operator == 'Div' and max(type1, type2) < 2:
-            max_bytes = max(dtype1.bytes, dtype2.bytes)
-            if type1 == type2 and type1 == 0:  # Unsigned integers
-                result_type = eval('dace.uint{}'.format(4 * max_bytes))
+    elif len(arguments) == 2:  # Binary operators
+
+        type1 = coarse_types[0]
+        type2 = coarse_types[1]
+        dtype1 = datatypes[0]
+        dtype2 = datatypes[1]
+        left_cast = None
+        right_cast = None
+
+        if _is_op_arithmetic(operator):
+
+            # Float/True division between integers
+            if operator == 'Div' and max(type1, type2) < 2:
+                # TODO: Leaving this here in case we implement a C/C++ flag
+                # max_bytes = max(dtype1.bytes, dtype2.bytes)
+                # if type1 == type2 and type1 == 0:  # Unsigned integers
+                #     result_type = eval('dace.uint{}'.format(8 * max_bytes))
+                # else:
+                #     result_type = eval('dace.int{}'.format(8 * max_bytes))
+                result_type = dace.float64
+                left_cast = dace.float64
+                right_cast = dace.float64
```

Comment:
we do have a flag like that - `Config.get_bool('compiler', 'default_data_types')`

Reply:
No, this flag is about how to interpret vague types like int, float, complex. This hypothetical flag would change operator behavior. For example '/' between integers is integer division in C/C++, but it is float division in Python.